### PR TITLE
Revert text_positions in Qwen25VL

### LINKED ISSUE #40136: Qwen2.5-VL-7B-Instruct: Significant accuracy regression on MMMU benchmark with transformers >=4.54.0
## Description

We've observed a significant accuracy regression when evaluating Qwen2.5-VL-7B-Instruct on the MMMU Literature benchmark after upgrading from transformers 4.53.3 to 4.54.0. The accuracy drops from **86.67%** to **73.33%**, representing a **13.34 percentage point decrease** (~15.4% relative regression).

## Environment

- **Model**: Qwen/Qwen2.5-VL-7B-Instruct
- **Evaluation framework**: lm-eval
- **Task**: mmmu_val_literature (30 samples)
- **Hardware**: CUDA-enabled GPU
- **Python**: 3.10
- **Other dependencies**: torch, torchvision, accelerate, Pillow

### Transformers 4.53.3 Output
```
hf-multimodal (pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True), gen_kwargs: (None), limit: None, num_fewshot: 0, batch_size: 8
|  Tasks   |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|----------|------:|------|-----:|------|---|-----:|---|-----:|
|Literature|      0|none  |     0|acc   |↑  |0.8667|±  |0.0631|
```

### Transformers 4.54.0 Output
```
hf-multimodal (pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True), gen_kwargs: (None), limit: None, num_fewshot: 0, batch_size: 8
|  Tasks   |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|----------|------:|------|-----:|------|---|-----:|---|-----:|
|Literature|      0|none  |     0|acc   |↑  |0.7333|±  |0.0821|
```

## Summary

| Version | Accuracy | Stderr | Difference from baseline |
|---------|----------|--------|-------------------------|
| 4.53.3  | 0.8667   | ±0.0631| -                       |
| 4.54.0  | 0.7333   | ±0.0821| -0.1334 (-15.4%)        |

## Possible causes

The regression may be related to:
- Changes in the default image processor (slow → fast) 
- Changes in the vision model loading mechanism
- Modifications to the multimodal pipeline

## Impact

This regression affects production systems using Qwen2.5-VL models for visual question answering and multimodal understanding tasks. Users upgrading to 4.54.0 may experience unexpected accuracy drops.

## Additional context

- The evaluation uses a fixed random seed for reproducibility
- Both tests were run on the same hardware with identical configurations
- The regression is statistically significant given the standard errors

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### 1. Install dependencies
```bash
pip install lm-eval torch torchvision accelerate Pillow
```

### 2. Test with transformers 4.53.3 (baseline)
```bash
pip install transformers==4.53.3

lm_eval \
    --model hf-multimodal \
    --model_args "pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True" \
    --tasks mmmu_val_literature \
    --num_fewshot 0 \
    --batch_size 8 \
    --verbosity INFO
```

### 3. Test with transformers 4.54.0
```bash
pip install transformers==4.54.0

lm_eval \
    --model hf-multimodal \
    --model_args "pretrained=Qwen/Qwen2.5-VL-7B-Instruct,dtype=bfloat16,add_bos_token=True,convert_img_format=True" \
    --tasks mmmu_val_literature \
    --num_fewshot 0 \
    --batch_size 8 \
    --verbosity INFO
```

## Expected behavior

The model accuracy should remain consistent or improve between minor version updates. A 15% relative performance regression on a standard benchmark is unexpected and concerning.

### LINKED ISSUE #40154: Qwen2.5VL is broken!
### System Info

- `transformers` version: 4.55.0
- Platform: Linux-5.15.0-1032-oracle-x86_64-with-glibc2.35
- Python version: 3.12.3
- Huggingface_hub version: 0.34.3
- Safetensors version: 0.6.1
- Accelerate version: 1.10.0
- Accelerate config:    not found
- DeepSpeed version: 0.17.4
- PyTorch version (accelerator?): 2.7.1+cu128 (CUDA)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: <fill in>
- Using GPU in script?: <fill in>
- GPU type: NVIDIA A100-SXM4-80GB

### Who can help?

_No response_

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

In recent PR: https://github.com/huggingface/transformers/pull/39447

A text position embedding is introduced:

https://github.com/huggingface/transformers/pull/39447/files#diff-72d7d1080bd2438bf28cf67cd035b5ef2a5b96beed7e947994fd5dbca55a0dbeR1624-R1628

```python
        if position_ids.ndim == 3 and position_ids.shape[0] == 4:
            text_position_ids = position_ids[0]
            position_ids = position_ids[1:]
        else:
            text_position_ids = position_ids[0]
```

The `text_position_ids` will be fed to the decoder layer: https://github.com/huggingface/transformers/pull/39447/files#diff-72d7d1080bd2438bf28cf67cd035b5ef2a5b96beed7e947994fd5dbca55a0dbeR1665

```python
            layer_outputs = decoder_layer(
                hidden_states,
                attention_mask=causal_mask_mapping[decoder_layer.attention_type],
                position_ids=text_position_ids,
                past_key_value=past_key_values,
                output_attentions=output_attentions,
                use_cache=use_cache,
```

OK. So far we have learned:

1. If `position_ids` inputs as a 3-dim tensor with shape (4, B, seq_len), then `text_position_ids` is the first (1, B, seq_len).
2. Otherwise, it just copy the first `position_ids[0]`

Now, you can easily verify that this value in training time is a 3-dim tensor with shape (3, B, seq_len), as shown in this figure: (note that I have visual input so seq_len != max(pos_id)

Image

Therefore, text_position_ids is simply (0, ..., 631) as you would expect in previous version.

However, in generation time, the text_position_ids is prepared by:

https://github.com/huggingface/transformers/pull/39447/files#diff-41bc5f79048d22eb30688e5440a89f4303ed02aa3d7c99aa807e3e7243e1fc0aR861-R864

```python
            if "position_ids" not in model_inputs:
                text_positions = torch.arange(input_ids, device=input_ids.device)[None, None, :]
            else:
                text_positions = model_inputs["position_ids"][None, ...]
```

If we take a look ahead, model_inputs["position_ids"] is prepared by GenerationMixin. In both case, text_positions is default to the (0, ..., seq_len-1). Then, this text_positions (which is already WRONG because it doesn't take care of the offset caused by potential visual tokens) will be concat to form a 4D position_ids.

```python
model_inputs["position_ids"] = torch.cat([text_positions, vision_positions], dim=0)
```

Then, such a value will be sent to decoder layer.

In summary, in training time, the decoder layer will see position_ids to be offsetted by visual token. But in generation, it see a plain (0,...,seq_len-1) position_ids.

This might be the root cause to https://github.com/huggingface/transformers/issues/40136

### Expected behavior

I expect it works normally.

### PR DESCRIPTION
**Pull request by pengzhenghao**
# What does this PR do?

Fixes # (issue)

Remove problematic text_position_ids preparation when generating with Qwen2.5VL model. 

Fixes #40154 (The Qwen 2.5 VL text position ID issue)
Fixes #40136 (Qwen2.5VL performance drop)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by pengzhenghao** [comment]
Note that the outreach of the issue might not limited to Qwen2.5VL. But I am not familiar with other model so the only thing I do here is to initialize the investigation on this issue.  @rahul-tuli let's see if this can get performance back.

**Comment by vasqu** [comment]
You only modified the modular file, but you need to apply the changes to the modeling file as well, i.e. `python utils/modular_converter qwen2_5_vl` or the sort (didnt check for exact naming).

**Comment by vasqu** [comment]
Thanks for investigating, it helps a lot! We are going to include #40161 in a patch for safety but for final fixes (maybe including this) will have to wait for 4.56.x potentially.

Let's see whether the original author of the mmmu issue can chime in and give feedback

**Comment by PavloFesenko** [comment]
@pengzhenghao Could you please add the issue numbers to `Fixes #<issue number>` in the PR description? The word `Fixes` tells GitHub that this PR will fix the underlying issue, displays a PR icon near the issue in the issue list, and automatically closes the issue once the PR is merged. 🙏

I guess that it should be OK to write something like this:
```
Fixes #40154 (The Qwen 2.5 VL text position ID issue)
Fixes #40136 (Qwen2.5VL performance drop)
```

**Review by zucchini-nlp — COMMENTED** [review]
Thanks for digging into the issue. I think we should not be deleting textual position ids here which is still needed for generation with FA2. Just removing text position will not solve it, because the root of issue is in https://github.com/huggingface/transformers/blob/cf58d703f9f6eb9d6295ce10354d24648e44e09a/src/transformers/models/qwen2_5_vl/modeling_qwen2_5_vl.py#L872-L898
We need to allow packed sequence with FA2 which was requested several times by the community, And instead define that `text_pos_ids = None` if the position ids do not contain 4 position types in the first dimension

**Comment by zucchini-nlp** [comment]
Can you also check if these tests are still passing on the models?

https://github.com/huggingface/transformers/blob/2914ceca202fd3a68a79fa88d03cf5b0435ec4f1/tests/test_modeling_common.py#L4265-L4284

**Comment by zucchini-nlp** [comment]
> Note that the outreach of the issue might not limited to Qwen2.5VL

You need to update all Qwen-VL family including Qwen-Omni, the issue is relevant to models using 3D RoPE

**Comment by gpantaz** [comment]
Hi!

Are there any updates on this? I am trying to finetune qwenvl-2.5 and I am noticing significant variations in the loss when using the latest transformers version (`4.55.4`). Within a couple of steps my loss is around 30-35 which seems suspicious. I also tried building from source using the forked library for this PR but I am seeing similar curves.

I tried a few previous versions and the latest one that seems to yield reasonable loss during each step is `4.49.0` I wonder if anyone else has experienced a similar behavior and if its related to this PR.



Thanks, 
George

**Comment by zucchini-nlp** [comment]
@pengzhenghao hey, do you still want to merge this PR? I can take it up, if you are busy

**Comment by zucchini-nlp** [comment]
I will assume you are busy and take over the issue tomorrow, because Qwen-VL models have high usage

**Comment by pengzhenghao** [comment]
Thank you so much for following this issue! @zucchini-nlp   I am indeed busy and also I am not familiar with other Qwen-VL models & FA2 impl. 

That would be great if you can take over!


> We need to allow packed sequence with FA2 which was requested several times by the community, And instead define that text_pos_ids = None if the position ids do not contain 4 position types in the first dimension

I am a little confused here. 

image

Seems like you are suggesting:

1. `text_position_ids` will be used as packed sequence and it should be [0, 1, ..., seq_len-1]
2. Therefore, for position_ids.shape[0]==4, `text_position_ids = position_ids[0]` is intended.
3. We instead should fix the problem that `text_position_ids=position_ids[0]` when input pos IDs are 3D, we should set `text_position_ids = None` in this case.

So, I think this naturally implies:

4. We should pass in a 3D position_ids when using Qwen-VL series models.

Do I understand correctly?

**Comment by zucchini-nlp** [comment]
@pengzhenghao yeah right, since we don't want to enforce all users to use packed sequence with FA2 we can just set `text_positions=None`. That way we will keep supporting packed sequences for users who requested it and fix the bug in the linked issue

I will check it tomorrow, and we definitely need more tests in Qwen given its specific RoPE type

**Comment by zucchini-nlp** [comment]
@pengzhenghao I located the bug and it is not in using packed or text position ids. The bug was actually in how position ids are computed when generating which was not same as computing in `forward`. I fixed it in https://github.com/huggingface/transformers/pull/40490 and checked with MMEval

Thanks for your PR and investigating it
